File: parser/src/main/scala/ru/mipt/acsl/decode/parser/ModelRegistry.scala
package ru.mipt.acsl.decode.parser

import java.nio.charset.StandardCharsets._

import com.google.common.io.Resources
import com.typesafe.scalalogging.LazyLogging
import ru.mipt.acsl.decode.model.registry.Registry
import ru.mipt.acsl.modeling.ErrorLevel

/**
  * @author Artem Shein
  */
object ModelRegistry extends LazyLogging {

  val Provider = new DecodeSourceProvider()
  val Config = new DecodeSourceProviderConfiguration("ru/mipt/acsl/decode")

  val RuntimeSource = SourceFileName("runtime")
  val Sources = Seq(RuntimeSource)

  def sourceName(nameWithoutExt: SourceFileName): String = nameWithoutExt.toFileNameWithoutExt + ".decode"

  def sourceResourcePath(nameWithoutExt: SourceFileName): String = "ru/mipt/acsl/decode/" + sourceName(nameWithoutExt)

  def sourceContents(nameWithoutExt: SourceFileName): String =
    Resources.toString(Resources.getResource(sourceResourcePath(nameWithoutExt)), UTF_8)

  def registryForSourceNames(sourceNames: Seq[SourceFileName]): Registry = registry(sourceNames.map(sourceContents))

  def registry(sources: Seq[String]): Registry = {
    val registry = Provider.provide(Config, sources)
    val resolvingResult = registry.resolve()
    if (resolvingResult.exists(_.level == ErrorLevel))
      resolvingResult.foreach(msg => logger.error(msg.text))
    registry
  }

  case class SourceFileName(name: String) {

    def toFileNameWithoutExt: String = name

    override def toString: String = toFileNameWithoutExt

  }

}

File: parser/src/main/java/ru/mipt/acsl/decode/parser/ModelRegistry.java
package ru.mipt.acsl.decode.parser;

import com.google.common.collect.Lists;
import com.google.common.io.Resources;
import ru.mipt.acsl.decode.model.proxy.ResolvingMessages;
import ru.mipt.acsl.decode.model.registry.Registry;
import ru.mipt.acsl.decode.model.registry.RegistryUtils;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.stream.Collectors;

/**
 * @author Artem Shein
 */
public class ModelRegistry {

    public static final DecodeSourceProvider Provider = new DecodeSourceProvider();
    public static final DecodeSourceProviderConfiguration Config = new DecodeSourceProviderConfiguration("mcc/decode");

    public static class Sources {

        public static final SourceFileName RUNTIME = SourceFileName.newInstance("runtime");
        public static final List<SourceFileName> ALL = Lists.newArrayList(RUNTIME);

        // prevent from instantiation
        private Sources() {

        }
    }

    public static String sourceName(SourceFileName nameWithoutExt) {
        return nameWithoutExt.fileNameWithoutExt() + ".decode";
    }

    public static String sourceResourcePath(SourceFileName nameWithoutExt) {
        return "mcc/" + sourceName(nameWithoutExt);
    }

    public static String sourceContents(SourceFileName nameWithoutExt) {
        try {
            return Resources.toString(Resources.getResource(sourceResourcePath(nameWithoutExt)), StandardCharsets.UTF_8);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    public static Registry registryForSourceNames(List<SourceFileName> sourceNames) {
        return registry(sourceNames.stream().map(ModelRegistry::sourceContents).collect(Collectors.toList()));
    }

    public static Registry registry(List<String> sources) {
        Registry registry = Provider.provide(Config, sources);
        ResolvingMessages resolvingResult = RegistryUtils.resolve(registry);
        if (resolvingResult.hasError())
            resolvingResult.messages().forEach(msg -> System.out.println(msg.text()));
        return registry;
    }

    // prevent from instantiation
    private ModelRegistry() {

    }

}
